File: Valid Parentheses/Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        
        queue = []

        for ch in s:
            match ch:
                case "(":
                    queue.append("(")

                case ")":
                    if len(queue) < 1:
                        return False
                    elif queue.pop() != "(": return False

                case "[":
                    queue.append("[")

                case "]":
                    if len(queue) < 1:
                        return False
                    elif queue.pop() != "[": return False

                case "{":
                    queue.append("{")

                case "}":
                    if len(queue) < 1:
                        return False
                    elif queue.pop() != "{": return False

        return not len(queue)

File: Valid Parentheses/Valid Parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    const pairs = {
        '(': ')',
        '[': ']',
        '{': '}'
    }

    const stack = [];

    // iterate through the string s
    for(const ch of s) {

        // ch is an opener
        if ( pairs[ch] ) stack.push(ch);

        // ch is a closer
        else if ( ch != pairs[ stack.pop() ] ) return false; // closer doesn't match opener!
    }

    // s is valid if no openers are left in the stack
    return !stack.length;  
};